### ISSUE TITLE
Terraform does not handle dependencies on count based resouces

### ISSUE THREAD

**Issue by cchildress** [hashicorp/terraform#16486]
### Terraform Version
0.10.7

### Terraform Configuration Files
```hcl
variable "availability_zones" {
  type = "map"

  default = {
    "0" = "us-west-2a"
    "1" = "us-west-2b"
    "2" = "us-west-2c"
  }
}

variable "foo_subnet_suffix" {
  type = "map"

  default = {
    "0" = ".0.0/24"
    "1" = ".1.0/24"
    "2" = ".2.0/24"
  }
}

data "aws_ami" "ubuntu_1604_lts_hvm" {
  most_recent = true

  filter {
    name   = "image-id"
    values = ["ami-8803e0f0"]
  }
}

resource "aws_vpc" "vpc" {
  cidr_block                       = "198.18.0.0/16"
}

resource "aws_subnet" "foo" {
  count             = "3"
  vpc_id            = "${aws_vpc.vpc.id}"
  availability_zone = "${lookup(var.availability_zones, count.index)}"
  cidr_block        = "$198.18${lookup(var.foo_subnet_suffix, count.index)}"
}

resource "aws_instance" "node" {
  ami = "${data.aws_ami.ubuntu_1604_lts_hvm.id}"

  instance_type               = "t2.large"
  availability_zone           = "${lookup(var.availability_zones, 2)}"
  subnet_id                   = "${aws_subnet.foo.2.id}"
}
```

### Debug Output
```
Error running plan: 1 error(s) occurred:

* aws_instance.node: Resource 'aws_subnet.foo' not found for variable 'aws_subnet.foo.2.id'
```

### Expected Behavior
Terraform should handle the dependency and create the count based resource that is a dependency first and then the other resource that has the dependency.

If the count number and the index numbers referenced in the downstream resources don't line up for some reason maybe that could become a fatal error during the run? Or try to be clever with comparing the count number with resources that try to depend on it?

### Actual Behavior
Terraform can't depend on a resource with a number in the resource ID because it hasn't been created yet. This forces multiple Terraform runs to do one build.

### Steps to Reproduce
1. Write a resource (like a subnet) with a count line specified.
2. Write a resource that depends on the counted resource (so like `"${aws_subnet.foo.1.id}"`
3. Run `terraform plan`

**Comment by apparentlymart** [comment]
Hi @cchildress! Sorry this isn't behaving as expected.

The specific error message here doesn't seem right, so there is indeed some sort of bug here, but at this time the syntax you used there isn't the correct way to write this reference. Instead, it should be written like this:

```hcl
subnet_id                   = "${aws_subnet.foo.*.id[2]}"
```

The expression `aws_subnet.foo.*.id` creates a list of ids and then the `[2]` takes the second element from this list.

The syntax you used here will be supported as part of the configuration language improvements we are currently working on, which will (among other changes) make `count`ed resources actually be a list of objects, whereas today they are not.

**Comment by cchildress** [comment]
Hey @apparentlymart 
Oh, well that would explain it... I'll test out the proper syntax once I get back to the office next week.

I did find the syntax you're describing in the docs, but only under how to use a template file. Maybe we could put a note about this in the doc that describes how to use count for a resource?

**Comment by cchildress** [comment]
Hi @apparentlymart 
I put in a PR to clarify this syntax in the docs if you want to take a look. #16640

**Comment by cchildress** [comment]
Fixed in #16640
